src/main/core/scheduler/scheduler.rs

```diff
@@ -0,0 +1,200 @@
+
+        // collect all of the hosts from the threads
+        self.pool.scope(|s| {
+            s.run(|t| {
+                THREAD_HOST.with(|x| {
+                    let host = x.borrow_mut().take().unwrap();
+                    *hosts[t.thread_idx].lock().unwrap() = Some(host);
+                });
+            });
+        });
+
+        // need to unref the host from the main thread so that the global allocation counter will be
+        // correctly updated
+        for host in hosts {
+            if let Some(host) = host.lock().unwrap().take() {
+                unsafe { crate::cshadow::host_unref(host.chost()) };
+            }
+        }
+
+        self.pool.join();
+    }
+}
+
+/// A wrapper around the work pool's scoped runner.
+pub struct SchedScope<'sched, 'pool, 'scope>
+where
+    'sched: 'scope,
+{
+    runner: TaskRunner<'pool, 'scope>,
+    // TODO: other schedulers (such as a thread-per-core scheduler) may need this, so leaving this
+    // here for now in case we need to provide an interface that has a 'sched lifetime
+    marker: PhantomData<&'sched Host>,
+}
+
+impl<'sched, 'pool, 'scope> SchedScope<'sched, 'pool, 'scope> {
+    /// Run the closure on all threads. The closure is given an index of the currently running
+    /// thread.
+    pub fn run(self, f: impl Fn(usize) + Sync + Send + 'scope) {
+        self.runner.run(move |task_context| {
+            // update the thread-local core affinity
+            if let Some(cpu_id) = task_context.cpu_id {
+                CORE_AFFINITY.with(|x| *x.borrow_mut() = Some(cpu_id));
+            }
+
+            (f)(task_context.thread_idx)
+        });
+    }
+
+    /// Run the closure on all threads. The closure is given an index of the currently running
+    /// thread and a host iterator.
+    ///
+    /// You must iterate over the provided `HostIter` to completion (until `next()` returns `None`),
+    /// otherwise this may panic. The host iterator is not a real [`std::iter::Iterator`], but
+    /// rather a fake iterator that behaves like a streaming iterator.
+    pub fn run_with_hosts(self, f: impl Fn(usize, &mut HostIter) + Send + Sync + 'scope) {
+        self.runner.run(move |task_context| {
+            // update the thread-local core affinity
+            if let Some(cpu_id) = task_context.cpu_id {
+                CORE_AFFINITY.with(|x| *x.borrow_mut() = Some(cpu_id));
+            }
+
+            THREAD_HOST.with(|host| {
+                let mut host = host.borrow_mut();
+
+                let mut host_iter = HostIter {
+                    host: Some(host.as_mut().unwrap()),
+                };
+
+                f(task_context.thread_idx, &mut host_iter);
+            });
+        });
+    }
+
+    /// Run the closure on all threads. The closure is given an index of the currently running
+    /// thread, a host iterator, and an element of `data`.
+    ///
+    /// You must iterate over the provided `HostIter` to completion (until `next()` returns `None`),
```

Comment:
"You must" -> "The passed function f must"

Reply:
Fixed.